src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,352 @@
+
+    internal static int WriteMetricsData(byte[] buffer, int writePosition, in Resources.Resource? resource, in Batch<Metric> batch)
+    {
+        foreach (var metric in batch)
+        {
+            var metricName = metric.MeterName;
+            if (!ScopeMetricsList.TryGetValue(metricName, out var metrics))
+            {
+                metrics = MetricListPool.Count > 0 ? MetricListPool.Pop() : new List<Metric>();
+                ScopeMetricsList[metricName] = metrics;
+            }
+
+            metrics.Add(metric);
+        }
+
+        writePosition = WriteResourceMetrics(buffer, writePosition, resource, ScopeMetricsList);
+        ReturnMetricListToPool();
+
+        return writePosition;
+    }
+
+    private static void ReturnMetricListToPool()
+    {
+        if (ScopeMetricsList.Count != 0)
+        {
+            foreach (var entry in ScopeMetricsList)
+            {
+                entry.Value.Clear();
+                MetricListPool.Push(entry.Value);
+            }
+
+            ScopeMetricsList.Clear();
+        }
+    }
+
+    private static int WriteResourceMetrics(byte[] buffer, int writePosition, in Resources.Resource? resource, in Dictionary<string, List<Metric>> scopeMetrics)
+    {
+        writePosition = ProtobufOtlpResourceSerializer.WriteResource(buffer, writePosition, resource);
+        writePosition = WriteScopeMetrics(buffer, writePosition, scopeMetrics);
+
+        return writePosition;
+    }
+
+    private static int WriteScopeMetrics(byte[] buffer, int writePosition, in Dictionary<string, List<Metric>> scopeMetrics)
+    {
+        if (scopeMetrics != null)
+        {
+            foreach (KeyValuePair<string, List<Metric>> entry in scopeMetrics)
+            {
+                writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.ResourceMetrics_Scope_Metrics, ProtobufWireType.LEN);
+                int resourceMetricsScopeMetricsLengthPosition = writePosition;
+                writePosition += ReserveSizeForLength;
+
+                writePosition = WriteScopeMetric(buffer, writePosition, entry.Value);
+
+                ProtobufSerializer.WriteReservedLength(buffer, resourceMetricsScopeMetricsLengthPosition, writePosition - (resourceMetricsScopeMetricsLengthPosition + ReserveSizeForLength));
+            }
+        }
+
+        return writePosition;
+    }
+
+    private static int WriteScopeMetric(byte[] buffer, int writePosition, in List<Metric> metrics)
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.ScopeMetrics_Scope, ProtobufWireType.LEN);
+        int instrumentationScopeLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        Debug.Assert(metrics.Any(), "Metrics collection is not expected to be empty.");
+        var meterName = metrics[0].MeterName;

```

Comment:
Should this be read every time or better to pass as parameters?

Reply:
These aren't available in the calling block, so can't be passed in as a parameter.